src/platforms/common/swdptap.c

```diff
@@ -89,31 +99,49 @@ static uint32_t swdptap_seq_in_clk_delay(size_t clock_cycles) __attribute__((opt
 static uint32_t swdptap_seq_in_clk_delay(const size_t clock_cycles)
 {
 	uint32_t value = 0;
-	for (size_t cycle = 0; cycle < clock_cycles; ++cycle) {
-		gpio_clear(SWCLK_PORT, SWCLK_PIN);
-		value |= gpio_get(SWDIO_IN_PORT, SWDIO_IN_PIN) ? 1U << cycle : 0U;
+	if (!clock_cycles)
+		return 0;
+	for (size_t cycle = clock_cycles; cycle--;) {
 		for (volatile uint32_t counter = target_clk_divider; counter > 0; --counter)
 			continue;
+		bool bit = gpio_get(SWDIO_IN_PORT, SWDIO_IN_PIN);
 		gpio_set(SWCLK_PORT, SWCLK_PIN);
 		for (volatile uint32_t counter = target_clk_divider; counter > 0; --counter)
 			continue;
+		value >>= 1U;
+		value |= (uint32_t)bit << 31U;
+		/* Reordering barrier */
+		__asm__("" ::: "memory");
+		gpio_clear(SWCLK_PORT, SWCLK_PIN);
+		/* Reordering barrier */
+		__asm__("" ::: "memory");
 	}
-	gpio_clear(SWCLK_PORT, SWCLK_PIN);
+	value >>= (32U - clock_cycles);
 	return value;
 }
 
 static uint32_t swdptap_seq_in_no_delay(size_t clock_cycles) __attribute__((optimize(3)));
 
 static uint32_t swdptap_seq_in_no_delay(const size_t clock_cycles)
 {
+	if (!clock_cycles)
```

Comment:
Same comment in this function as the one above about count-down vs count-up loops.